src/controllers/resource-store/index.tsx

```diff
@@ -104,75 +104,99 @@ export const actions: Actions = {
   }): Promise<RouteResourceResponse<unknown>> => {
     const { type, getKey, getData, maxAge } = resource;
     const { prefetch, timeout } = options;
-    const { data: resourceStoreData, context } = getState();
+    const { context, ...resourceStoreState } = getState();
     const key = getKey(routerStoreContext, context);
-    const slice = getSliceForResource(
-      { data: resourceStoreData },
-      { type, key }
-    );
+    const originalSlice = getSliceForResource(resourceStoreState, {
+      type,
+      key,
+    });
 
-    if (slice.loading) {
-      return slice;
+    if (originalSlice.loading) {
+      return originalSlice;
     }
 
     dispatch(validateLRUCache(resource, key));
 
-    const pending = {
-      ...slice,
-      data: maxAge === 0 ? null : slice.data,
-      error: maxAge === 0 ? null : slice.error,
+    const promise = getData(
+      { ...routerStoreContext, isPrefetch: !!prefetch },
+      context
+    );
+    const pendingSlice = {
+      ...originalSlice,
+      data: maxAge === 0 ? null : originalSlice.data,
+      error: maxAge === 0 ? null : originalSlice.error,
       loading: true,
-      promise: getData(
-        { ...routerStoreContext, isPrefetch: !!prefetch },
-        context
-      ),
+      promise,
       accessedAt: getAccessedAt(),
     };
 
-    dispatch(setResourceState(type, key, pending));
-
-    const response = {
-      ...pending,
-    };
+    dispatch(setResourceState(type, key, pendingSlice));
 
+    // do not introduce additional awaited function call here, it will cause additional microtasks
+    let resolvedSlice;
     try {
-      response.error = null;
-
       if (timeout) {
         const timeoutGuard = generateTimeGuard(timeout);
-        const maybeData = await Promise.race([
-          pending.promise,
-          timeoutGuard.promise,
-        ]);
+        const maybeData = await Promise.race([promise, timeoutGuard.promise]);
 
         if (!timeoutGuard.isPending) {
-          response.data = null;
-          response.error = new TimeoutError(type);
-          response.loading = true;
-          response.promise = null;
+          // timeout state intentionally looks like hung loading, with exception of the Error
+          // however promise must be null to avoid actually hanging
+          resolvedSlice = {
+            data: null,
+            error: new TimeoutError(type),
+            loading: true,
+            promise: null,
+          };
         } else {
           timeoutGuard.timerId && clearTimeout(timeoutGuard.timerId);
-          response.data = maybeData;
-          response.loading = false;
+          resolvedSlice = {
+            data: maybeData,
+            error: null, // previous error is removed
+            loading: false,
+            promise,
```

Comment:
`pendingSlice` already has `promise` being set. Is there any value setting it to `resolvedSlice` ?

Reply:
IMO yes. In this block there is a branch where a timeout will produce `promise: null`. Without declaring the value of `promise` in each branch it is harder to compare the final state at a glance.

Reply:
Then, should we go all explicit and no longer spread `pendingSlice` in `finalSlice` ? As I'd say either we leverage spreading for all non changing data, or we explicit set the state for every step. Seems like the latter is preferable for clarity, so I'd enforce it